test/cppsuite/test_harness/workload/database_operation.h

```diff
@@ -52,161 +57,279 @@ class database_operation {
+                    std::to_string(i),
+                  DEBUG_TRACE);
+                thread_collections.push_back(collection_names[j]);
+            }
+            tm.add_thread(
+              populate, i, thread_collections, connection_manager::instance().create_session());
         }
-        debug_print("Populate stage done", DEBUG_TRACE);
+
+        /* Wait for our populate threads to finish and then join them. */
+        debug_print("Populate: waiting for threads to complete.", DEBUG_INFO);
+        tm.join();
+
+        debug_print("Populate: finished.", DEBUG_INFO);
+    }
+
+    /* Basic insert operation that adds a new key every rate tick. */
+    virtual void
+    insert_operation(thread_context *tc)
+    {
+        debug_print(
+          type_string(tc->type) + " thread {" + std::to_string(tc->id) + "} commencing loop.",
+          DEBUG_INFO);
     }
 
     /* Basic read operation that walks a cursors across all collections. */
     virtual void
     read_operation(thread_context *tc)
     {
+        debug_print(
+          type_string(tc->type) + " thread {" + std::to_string(tc->id) + "} commencing loop.",
+          DEBUG_INFO);
         WT_CURSOR *cursor;
         std::vector<WT_CURSOR *> cursors;
 
         /* Get a cursor for each collection in collection_names. */
         for (const auto &it : tc->database.get_collection_names()) {
             testutil_check(tc->session->open_cursor(tc->session, it.c_str(), NULL, NULL, &cursor));
             cursors.push_back(cursor);
+            debug_print("Adding collection to read thread: " + it, DEBUG_TRACE);
         }
 
-        while (!cursors.empty() && tc->running()) {
+        while (tc->running()) {
             /* Walk each cursor. */
             for (const auto &it : cursors) {
                 if (it->next(it) != 0)
                     it->reset(it);
             }
+            tc->sleep();
         }
     }
 
     /*
-     * Basic update operation that updates all the keys to a random value in each collection.
+     * Basic update operation that uses a random cursor to update values in a randomly chosen
+     * collection.
      */
     virtual void
     update_operation(thread_context *tc)
     {
+        debug_print(
+          type_string(tc->type) + " thread {" + std::to_string(tc->id) + "} commencing loop.",
+          DEBUG_INFO);
         WT_CURSOR *cursor;
         WT_DECL_RET;
         wt_timestamp_t ts;
-        std::vector<WT_CURSOR *> cursors;
-        std::vector<std::string> collection_names = tc->database.get_collection_names();
+        /*
+         * We need three maps here, one to maintain our list of random cursors, one to maintain the
+         * list of update cursors. We can't update using a random cursor, as defined by the
+         * WiredTiger API.
+         *
+         * The final map provides a quick lookup for collection names.
+         */
+        std::map<uint64_t, WT_CURSOR *> random_cursors;
```

Comment:
Does it make sense to just have some intermediate struct like:
```
struct CollectionCursors {
  const std::string collection_name;
  WT_CURSOR *update_cursor, *random_cursor;
};

std::map<uint64_t, CollectionCursors> collection_cursors;
```

Then you can just do one lookup to get all the relevant info. Can even use `std::tuple` if you don't want to make an ad hoc struct.

Reply:
ee0baa630